src/main/java/com/yahoo/bullet/querying/evaluators/Evaluator.java

```diff
@@ -0,0 +1,108 @@
+        BINARY_OPERATORS.put(LazyExpression.Operation.MUL, BinaryOperations.MUL);
+        BINARY_OPERATORS.put(LazyExpression.Operation.DIV, BinaryOperations.DIV);
+        BINARY_OPERATORS.put(LazyExpression.Operation.EQUALS, BinaryOperations.EQUALS);
+        BINARY_OPERATORS.put(LazyExpression.Operation.NOT_EQUALS, BinaryOperations.NOT_EQUALS);
+        BINARY_OPERATORS.put(LazyExpression.Operation.GREATER_THAN, BinaryOperations.GREATER_THAN);
+        BINARY_OPERATORS.put(LazyExpression.Operation.LESS_THAN, BinaryOperations.LESS_THAN);
+        BINARY_OPERATORS.put(LazyExpression.Operation.GREATER_THAN_OR_EQUALS, BinaryOperations.GREATER_THAN_OR_EQUALS);
+        BINARY_OPERATORS.put(LazyExpression.Operation.LESS_THAN_OR_EQUALS, BinaryOperations.LESS_THAN_OR_EQUALS);
+        BINARY_OPERATORS.put(LazyExpression.Operation.REGEX_LIKE, BinaryOperations.REGEX_LIKE);
+        BINARY_OPERATORS.put(LazyExpression.Operation.SIZE_IS, BinaryOperations.SIZE_IS);
+        BINARY_OPERATORS.put(LazyExpression.Operation.CONTAINS_KEY, BinaryOperations.CONTAINS_KEY);
+        BINARY_OPERATORS.put(LazyExpression.Operation.CONTAINS_VALUE, BinaryOperations.CONTAINS_VALUE);
+        BINARY_OPERATORS.put(LazyExpression.Operation.AND, BinaryOperations.AND);
+        BINARY_OPERATORS.put(LazyExpression.Operation.OR, BinaryOperations.OR);
+        BINARY_OPERATORS.put(LazyExpression.Operation.XOR, BinaryOperations.XOR);
+        BINARY_OPERATORS.put(LazyExpression.Operation.FILTER, BinaryOperations.FILTER);
+        UNARY_OPERATORS.put(LazyExpression.Operation.NOT, UnaryOperations.NOT);
+        UNARY_OPERATORS.put(LazyExpression.Operation.SIZE_OF, UnaryOperations.SIZE_OF);
+    }
+
+    Evaluator(LazyExpression expression) {
+        this.type = expression.getType();
+    }
+
+    public abstract TypedObject evaluate(BulletRecord record);
+
+    protected TypedObject cast(TypedObject object) {
+        if (type == null || ((object.getType() == Type.LIST || object.getType() == Type.MAP) && object.getPrimitiveType() == type)) {
+            return object;
+        }
+        if (object.getType() == Type.LIST) {
+            List<Object> objects = object.getList();
+            return new TypedObject(Type.LIST, objects.stream().map(o -> TypedObject.typeCastFromObject(type, o).getValue()).collect(Collectors.toList()));
+        }
+        if (object.getType() == Type.MAP) {
+            Map<String, Object> map = object.getMap();
+            Map<String, Object> newMap = new HashMap<>();
+            map.forEach((key, value) -> newMap.put(key, TypedObject.typeCastFromObject(type, value).getValue()));
+            return new TypedObject(Type.MAP, newMap);
+        }
+        return object.forceCast(type);
+    }
+
+    /**
+     * What is polymorphism?
+     *
+     * @param expression
+     * @return
+     */
+    public static Evaluator build(LazyExpression expression) {
+        if (expression instanceof LazyNull) {
```

Comment:
Just make each Expression return you the right Evaluator?

Reply:
I thought we agreed to return the right class as a compromise, since at least the logic isn't there and we don't have to get the Java police on us for breaking polymorphism

Reply:
ok, I have no problems with that.